sdk/keyvault/azure-keyvault-keys/tests/_test_case_async.py

```diff
@@ -0,0 +1,52 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import os
+from urllib.parse import urlparse
+
+from azure.keyvault.keys.aio import KeyClient
+from azure.keyvault.keys.crypto.aio import CryptographyClient
+from azure.keyvault.keys._shared import HttpChallengeCache
+from parameterized import parameterized
+import pytest
+
+from _shared.test_case_async import KeyVaultTestCase
+
+
+def suffixed_test_name(testcase_func, param_num, param):
+    suffix = "mhsm" if param.kwargs.get("is_hsm") else "vault"
+    return "{}_{}".format(testcase_func.__name__, parameterized.to_safe_name(suffix))
+
+
+class KeysTestCase(KeyVaultTestCase):
```

Comment:
It would definitely be easy to create a mixin for these classes that contains `create_key_client`, `create_crypto_client`, and `_skip_if_not_configured`. Creating a base class that handles `setUp` and `tearDown` is probably doable as well, with these classes then inheriting from the base class and their respective KeyVaultTestCase. The latter would share code more efficiently but would be a bit more confusing -- is there a different setup that you had in mind, like a mixin for KeyVaultTestCase?

Reply:
I was thinking that the async test case classes could inherit KeysTestCase and mix in the async methods. I was neglecting the create_client methods though. They make it look more complex than it's worth but maybe I'm just impatient or shortsighted. Anyway, not a high priority.

Reply:
I just pushed a base class implementation that shares all the code, with some (but I think acceptable levels of) complexity in the client creation methods